### PR TITLE
Optimize array_distinct(array_sort(…))

### LINKED ISSUE #8777: Optimize array_distinct(array_sort(…))
The calls should be inverted, as sorting the distinct set is better if the set is smaller (and never worse).

### PR DESCRIPTION
**Pull request by mdesmet**
Fixes #8777 

As I'm using ArrayConstructor in my test, I had to add the transactionID in the test suite, however I think this broke some other tests. Need some advice on best approach to override the session only for this test or not require a transactionid.

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/scalar/ArrayDistinctFunction.java [hunk 1]
```diff
@@ -31,16 +31,18 @@
...
 import static io.trino.operator.aggregation.TypedSet.createDistinctTypedSet;
+import static io.trino.operator.scalar.ArrayDistinctFunction.*;
 import static io.trino.spi.function.InvocationConvention.InvocationArgumentConvention.BLOCK_POSITION;
 import static io.trino.spi.function.InvocationConvention.InvocationReturnConvention.FAIL_ON_NULL;
 import static io.trino.spi.function.OperatorType.HASH_CODE;
 import static io.trino.spi.function.OperatorType.IS_DISTINCT_FROM;
 import static io.trino.spi.type.BigintType.BIGINT;
 
-@ScalarFunction("array_distinct")
+@ScalarFunction(NAME)
```
Comment (hashhar) [inline core/trino-main/src/main/java/io/trino/operator/scalar/ArrayDistinctFunction.java:41]: revert this change. The `@ScalarFunction("<func-name>")` makes it easier to navigate the codebase IMO.

Comment (hashhar) [reply core/trino-main/src/main/java/io/trino/operator/scalar/ArrayDistinctFunction.java:41]: Same in ArraySortFunction.

**Comment by martint** [comment]
@cla-bot check

File: core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        public Visitor(Map<NodeRef<Expression>, Type> expressionTypes, Metadata metadata)
+        {
+            this.expressionTypes = ImmutableMap.copyOf(requireNonNull(expressionTypes, "expressionTypes is null"));
+            this.metadata = metadata;
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (node.getName().getSuffix().contains(String.format("@%s@", ArrayDistinctFunction.NAME))
```
Comment (martint) [inline core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java:74]: The content of `FunctionCall.getName()` is an opaque identifier for the underlying function resolved during analysis. Attempting to inspect that string is not the right way to do this. Instead, you need to get the name from the corresponding `ResolvedFunction` object, which you can obtain from `MetadataManager` (see how it's done in `DeterminismEvaluator`):

```
metadata.decodeFunction(node.getName())
```

Ideally, you'd just compare the id of the ResolvedFunction with the id of a well-known function, like it's done in `RewriteSpatialPartitioningAggregation.isStEnvelopeFunctionCall()` but that only works if you know the specific argument types in advance. So, checking the name should be ok for now.

Comment (mdesmet) [reply core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java:74]: Makes total sense. Will take care of it!

File: core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (metadata.getFunctionMetadata(metadata.decodeFunction(node.getName())).getActualName().equals(ArrayDistinctFunction.NAME)
+                    && node.getArguments().size() > 0
+                    && node.getArguments().get(0) instanceof FunctionCall
+                    && metadata.getFunctionMetadata(metadata.decodeFunction(((FunctionCall) node.getArguments().get(0)).getName())).getActualName().equals(ArraySortFunction.NAME)) {
+                FunctionCall sortFunctionCall = (FunctionCall) node.getArguments().get(0);
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java:78]: Call `treeRewriter.defaultRewrite()` on the `sortFunctionCall` to process the arguments recursively. Add test with nested rewrite.

File: core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            this.expressionTypes = ImmutableMap.copyOf(requireNonNull(expressionTypes, "expressionTypes is null"));
+            this.metadata = metadata;
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (metadata.getFunctionMetadata(metadata.decodeFunction(node.getName())).getActualName().equals(ArrayDistinctFunction.NAME)
+                    && node.getArguments().size() > 0
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java:75]: The `array_distinct` function should have exactly one argument. You could add a checkState here.
At least, change this condition to `node.getArguments().size() == 1` so that you bail out if there are additional arguments, because they're not handled in the following code.

File: core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+                    && node.getArguments().get(0) instanceof FunctionCall
+                    && metadata.getFunctionMetadata(metadata.decodeFunction(((FunctionCall) node.getArguments().get(0)).getName())).getActualName().equals(ArraySortFunction.NAME)) {
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java:77]: The name "array_sort" identifies two different functions. According to Trino docs, there are:
`array_sort(x) → array`
`array_sort(array(T), function(T, T, int)) -> array(T)`

Make sure that only the first argument goes to `array_distinct`, and the second argument, if present, remains in the `array_sort` call.
Add test for such case.

File: core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public final class ArraySortAfterArrayDistinctRewriter
+{
+    public static Expression rewrite(Expression expression, Map<NodeRef<Expression>, Type> expressionTypes, Metadata metadata)
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java:39]: Is this method used anywhere outside the method below? If not, please inline it.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+public class ArraySortAfterArrayDistinct
+        extends ExpressionRewriteRuleSet
+{
+    public ArraySortAfterArrayDistinct(Metadata metadata, TypeAnalyzer typeAnalyzer)
+    {
+        super(createRewrite(metadata, typeAnalyzer));
+    }
+
+    @Override
+    public Set<Rule<?>> rules()
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java:35]: Why do you override this method?

**Review by kasiafi — COMMENTED** [review]
Thanks for working on this!

The alternative way of testing the rewrite is to use the mechanism for testing Optimizer rules. See `TestPushCastIntoRow` as an example.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java (hunk 7 above)
Comment (mdesmet) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java:35]: I based myself on another code that was having these rules setup. I will test without this Override and see if I can get rid of that.

File: core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java (hunk 5 above)
Comment (mdesmet) [reply core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java]: Hi @kasiafi , I've implemented support for this, however i've not managed to get the test working (See the commented out test). The test is faling with java.lang.IllegalStateException: missing expression for alias x. If you would provide any pointers on how to solve that one, that would be helpfull. :)

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java (hunk 7 above)
Comment (mdesmet) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java:50]: Actually this is needed, as following code can't handle aggregations as it can't resolve functions like count (window functions):
` ResolvedFunction decodeFunction(QualifiedName name);`

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (isRequiredFunctionCall(node, ArrayDistinctFunction.NAME)
+                    && node.getArguments().size() > 0
+                    && node.getArguments().get(0) instanceof FunctionCall) {
+                FunctionCall arraySortFunctionCall = (FunctionCall) node.getArguments().get(0);
+                if (isRequiredFunctionCall(arraySortFunctionCall, ArraySortFunction.NAME)) {
+                    checkState(node.getArguments().size() == 1, "Function array_distinct called with unsupported number of arguments");
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java:94]: After a second glance, I think that we should check 
`node.getArguments().size() == 1` and `arraySortArguments.size() <= 2` and just bail out if they don't hold, instead of throwing error. This is not responsibility of this rewrite to validate function calls.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+                    FunctionCall rewritten = treeRewriter.defaultRewrite(arraySortFunctionCall, context);
+
+                    List<Expression> arraySortArguments = rewritten.getArguments();
+                    checkState(arraySortArguments.size() <= 2, "Function array_sort called with unsupported number of arguments");
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java:99]: Do the check before you call `defaultRewrite`.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+                    Map<NodeRef<Expression>, Type> expressionTypes = typeAnalyzer.getTypes(this.context.getSession(), this.context.getSymbolAllocator().getTypes(), rewritten);
+
+                    List<Type> arraySortArgumentsTypes = arraySortArguments.stream()
+                            .map(NodeRef::of)
+                            .map(expressionTypes::get)
+                            .collect(toImmutableList());
+
+                    FunctionCall arrayDistinctCall = new FunctionCallBuilder(metadata)
+                            .setName(QualifiedName.of(ArrayDistinctFunction.NAME))
+                            .setArguments(Lists.newArrayList(arraySortArgumentsTypes.get(0)), Lists.newArrayList(arraySortArguments.get(0)))
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java:110]: Use `ImmutableList.of(...)`.

File: core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java (hunk 5 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/planner/ArraySortAfterArrayDistinctRewriter.java]: Apparently, the assertions cannot handle a lambda.

I think that you could add this test case in `query`. Also, you could add some more examples there, with non-distinct and unsorted arrays as input, to show that the results are correct.

File: core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+//    // TODO: should figure out why this is failing, although the plan seems correct
+//    public void withLambda()
+//    {
+//        test("ARRAY_DISTINCT(ARRAY_SORT(ARRAY[1], (x, y) -> IF(x < y, -1, IF(x > y, 1, 0 ))))", "ARRAY_SORT(ARRAY_DISTINCT(ARRAY[1]), (x, y) -> IF(x < y, -1, IF(x > y, 1, 0 )))");
+//    }
+
+    @Test
+    public void withRecursion()
+    {
+        test("ARRAY_DISTINCT(ARRAY_SORT(ARRAY_DISTINCT(ARRAY_SORT(ARRAY['a']))))", "ARRAY_SORT(ARRAY_DISTINCT(ARRAY_SORT(ARRAY_DISTINCT(ARRAY['a']))))");
```
Comment (kasiafi) [inline core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java:49]: Here we have the outcome of single application of the rule.
If you added this test case to `TestLogicalPlanner`, we could see the function call chain fully sorted.

File: core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        extends BaseRuleTest
+{
+    @Test
+    public void basic()
+    {
+        test("ARRAY_DISTINCT(ARRAY_SORT(ARRAY['a']))", "ARRAY_SORT(ARRAY_DISTINCT(ARRAY['a']))");
+    }
+
+//    @Test
+//    // TODO: should figure out why this is failing, although the plan seems correct
```
Comment (kasiafi) [inline core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java:40]: Apparently, the assertions cannot handle the lambda. 

Please add this test case to `query`. Also, add some more examples there with unsorted and non-distinct arrays to show thet the results are correct.

Comment (mdesmet) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java:40]: Which test you actually mean with `query`?

File: core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java (hunk 11 above)
Comment (mdesmet) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java:49]: I need some pointers here.

I've tried to add some simple test here:

```
    @Test
    public void testArraySortAfterArrayDistinctRecursive()
    {
        assertPlan("SELECT ARRAY_SORT(items) from (VALUES (ARRAY ['cat', 'dog', 'dog'])) t(items)",
                output(project(ImmutableMap.of("output", expression("ARRAY_SORT(items)")),
                        values())));
    }
```

However this results in following plan:

```
Output[_col0]
│   Layout: [array_sort:array(varchar(3))]
│   _col0 := array_sort
└─ Values
       Layout: [array_sort:array(varchar(3))]
       ("$literal$"(from_base64('DgAAAFZBUklBQkxFX1dJRFRIAwAAAAMAAAAGAAAACQAAAAAJAAAAY2F0ZG9nZG9n')))

```

How can I express such a plan?

Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java:49]: In your example, first a VALUES-related optimization `MergeProjectWithValues` fires, and the `array_sort` call is inlined into VALUES, and then the expression in VALUES is evaluated into literal. To avoid this, try using a non-deterministic expression (it shouldn't be folded into a literal):
```
VALUES array_distinct(array_sort(array_distinct(array_sort(ARRAY[0e0, 1e0, rand()]))))
```

File: core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java (hunk 12 above)
Comment (mdesmet) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java:40]: Thanks, I actually figured out that one already :)

File: core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java (hunk 11 above)
Comment (mdesmet) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java:49]: Great, I will try it out!

Comment (mdesmet) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java:49]: Getting a bit further. However failing in ValuesMatcher as it seems the ARRAY_CONSTRUCTOR doesn't match up (failing at `!expectedRows.equals(valuesNode.getRows())`)

```
    @Test
    public void testArraySortAfterArrayDistinctRecursive2()
    {
        FunctionCall randomFunction = new FunctionCall(
                getQueryRunner().getMetadata().resolveFunction(QualifiedName.of("rand"), ImmutableList.of()).toQualifiedName(),
                ImmutableList.of());

        FunctionCall arrayConstructor = new FunctionCall(
                getQueryRunner().getMetadata().resolveFunction(QualifiedName.of("array_constructor"), fromTypes(DOUBLE)).toQualifiedName(),
                ImmutableList.of(new DoubleLiteral("0e0"), new DoubleLiteral("1e0"), randomFunction));

        assertPlan("SELECT * FROM (VALUES (ARRAY[0e0, 1e0, rand()])) T(x)",
                output(values(
                                ImmutableList.of("x"),
                                ImmutableList.of(ImmutableList.of(arrayConstructor)))));
    }
```

Comment (kasiafi) [reply core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java:49]: That's because the resolved function for `array_constructor` doesn't match. If you replace it with the below  (it has `fromTypes(DOUBLE, DOUBLE, DOUBLE)`), it passes.

```
FunctionCall arrayConstructor = new FunctionCall(
                getQueryRunner().getMetadata().resolveFunction(QualifiedName.of("array_constructor"), fromTypes(DOUBLE, DOUBLE, DOUBLE)).toQualifiedName(),
                ImmutableList.of(new DoubleLiteral("0e0"), new DoubleLiteral("1e0"), randomFunction));
```

**Comment by mdesmet** [comment]
@martint , @kasiafi: Any other things that need to be done to get this merged?

See also: 

https://github.com/trinodb/trino/blob/14e60285b3ef8f519ea67b1f4a54385f6bad6ebf/core/trino-main/src/main/java/io/trino/metadata/ResolvedFunction.java#L61

**Comment by kasiafi** [comment]
> Do the check before you call defaultRewrite.

I'm sorry, this was bad advice. Actually, to get the most from a single application of the rule, you should run the default rewrite first:
```
            FunctionCall rewritten = treeRewriter.defaultRewrite(node, context);
            if (isRequiredFunctionCall(rewritten, ArrayDistinctFunction.NAME) &&
                    rewritten.getArguments().size() == 1 &&
                    rewritten.getArguments().get(0) instanceof FunctionCall &&
                    isRequiredFunctionCall((FunctionCall) rewritten.getArguments().get(0), ArraySortFunction.NAME)) {
                // ...
            }
            return rewritten;
```

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+                        .build();
+
+                FunctionCallBuilder arraySortCallBuilder = FunctionCallBuilder.resolve(session, metadata)
+                        .setName(QualifiedName.of(ArraySortFunction.NAME))
+                        .addArgument(arraySortArgumentsTypes.get(0), arrayDistinctCall);
+
+                if (arraySortArguments.size() == 2) {
+                    arraySortCallBuilder.addArgument(arraySortArgumentsTypes.get(1), arraySortArguments.get(1));
+                }
+                return treeRewriter.defaultRewrite(arraySortCallBuilder.build(), context);
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java:104]: The rewrite is wrapped in an iterative rule, so it will be run again if there's something more to do. Let's not do the default rewrite twice in one invocation.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            FunctionCall rewritten = treeRewriter.defaultRewrite(node, context);
+            if (metadata.decodeFunction(rewritten.getName()).getSignature().getName().equals(ArrayDistinctFunction.NAME) &&
+                    rewritten.getArguments().size() == 1 &&
+                    rewritten.getArguments().get(0) instanceof FunctionCall &&
+                    metadata.decodeFunction(((FunctionCall) rewritten.getArguments().get(0)).getName()).getSignature().getName().equals(ArraySortFunction.NAME)) {
+                FunctionCall arraySortFunctionCall = (FunctionCall) rewritten.getArguments().get(0);
+                ResolvedFunction arraySortResolvedFunction = metadata.decodeFunction(arraySortFunctionCall.getName());
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java:85]: Please try to structure the code so as to call `metadata.decodeFunction` once for a Functioncall.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java (hunk 13 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java]: Yeah, I see that this test case needs multiple passes of the rewrite to fully sort the `array_distinct` and `array_sort` calls.
What I mean by "it's wrapped in an iterative rule" is that even if a single pass doesn't do the whole job, the rule should be called again, and again... until the call chain is fully sorted.

This test runs a single invocation of the rule, so I wouldn't expect the result fully sorted.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+            this.session = session;
+        }
+
+        @Override
+        public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            FunctionCall rewritten = treeRewriter.defaultRewrite(node, context);
+
+            if (metadata.decodeFunction(rewritten.getName()).getSignature().getName().equals(ArrayDistinctFunction.NAME) &&
+                    rewritten.getArguments().size() == 1 &&
```
[the remark this answers is not in the document]
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java:81]: Yes, you can use `getOnlyElement()` instead to get the argument.

File: core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java (hunk 13 above)
Comment (mdesmet) [reply core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ArraySortAfterArrayDistinct.java]: Thanks for the clarification, I looked at the IterativeOptimizer and indeed it will exhaustively go through a set of rules and try to apply them until. So i will adapt the test for one single invocation and remove the recursive call.

File: core/trino-main/src/test/java/io/trino/sql/query/TestArraySortAfterArrayDistinct.java [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+import org.testng.annotations.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestArraySortAfterArrayDistinct
+{
+    @Test
+    public void test()
+    {
+        try (QueryAssertions assertions = new QueryAssertions()) {
```
Comment (kasiafi) [inline core/trino-main/src/test/java/io/trino/sql/query/TestArraySortAfterArrayDistinct.java:25]: Could you extract initialization and teardown methods?
See `query/TestRecursiveCte.java` for reference.
Then you don't need to wrap each test in a `try` block.

Also, annotate the class with `@TestInstance(PER_CLASS)`.

File: core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestArraySortAfterArrayDistinct.java [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.Map;
+
+import static io.trino.sql.planner.assertions.PlanMatchPattern.project;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.values;
+
+public class TestArraySortAfterArrayDistinct
+        extends BaseRuleTest
+{
+    @Test
+    public void basic()
```
Comment (kasiafi) [inline core/trino-main/src/test/java/io/trino/sql/planner/iterative/rule/TestArraySortAfterArrayDistinct.java:34]: Rename the methods so that the names are more verbose, and they should begin with "test".

**Review by kasiafi — APPROVED** [review]
Looks good. Just a couple minor comments.

**Review by kasiafi — COMMENTED** [review]
Ping me for merge when you update the method name and pass tests.

**Comment by mdesmet** [comment]
@kasiafi: Change has been applied and build is green. Ready for merge!
